subsys/fs/zffs/src/zffs_dir.c

```diff
@@ -0,0 +1,699 @@
+		      u32_t len, void *data, const void *name)
+{
+	struct dir_compar_data *dcd = data;
+	int rc;
+
+	dcd->node_data->id = ((struct zffs_slist_node *)node)->id;
+	rc = zffs_misc_load_node(zffs, pointer, len, dcd->node_data,
+				  dcd->snode);
+	if (rc) {
+		return 1;
+	}
+
+	return strcmp(dcd->node_data->name, name) ? 1 : 0;
+}
+
+int zffs_dir_search(struct zffs_data *zffs,
+		     struct zffs_area_pointer *pointer,
+		     const struct zffs_dir *dir, const char *name,
+		     struct zffs_node_data *node_data, sys_snode_t **snode)
+{
+	struct zffs_slist_node node;
+	struct dir_compar_data data = {.node_data = node_data, .snode = snode};
+
+	return zffs_slist_search(zffs, *pointer, &dir->list, &node, &data,
+				  name, dir_compar);
+}
+
+int zffs_dir_search_for_node_data(struct zffs_data *zffs,
+				   struct zffs_area_pointer *pointer,
+				   struct zffs_node_data *node_data,
+				   const char *name, sys_snode_t **snode)
+{
+	int rc;
+
+	if (*name == '\0') {
+		struct zffs_area_pointer rp = *pointer;
+
+		rc = zffs_object_open(zffs, &rp, ZFFS_ROOT_ID, NULL);
+		if (rc < 0) {
+			return rc;
+		}
+
+		node_data->id = ZFFS_ROOT_ID;
+
+		rc = zffs_misc_load_node(zffs, &rp, rc, node_data, snode);
+	}
+	else if (node_data->type == ZFFS_TYPE_DIR) {
+		struct dir_compar_data data = {.node_data = node_data,
+					       .snode = snode};
+
+		struct zffs_slist list;
+		struct zffs_slist_node node;
+
+		zffs_slist_init(&list);
+
+		list.head = node_data->dir.head;
+
+		rc = zffs_slist_search(zffs, *pointer, &list, &node, &data,
+					name, dir_compar);
+	}
+	else {
+		rc = -ENOTDIR;
+	}
+
+	return rc;
+}
+
+int zffs_dir_open(struct zffs_data *zffs, struct zffs_area_pointer *pointer,
+		   const struct zffs_node_data *node_data,
+		   struct zffs_dir *dir)
+{
+	sys_snode_t *snode;
+
+	if (node_data->type != ZFFS_TYPE_DIR) {
+		return -ENOTDIR;
+	}
+
+	SYS_SLIST_FOR_EACH_NODE(&zffs->opened, snode)
+	{
+		if (((struct zffs_dir *)snode)->id == node_data->id) {
+			return -EBUSY;
+		}
+	}
+
+	dir->id = node_data->id;
+	dir->list.head = node_data->dir.head;
+
+	dir->list.wait_update = false;
+	dir->node.id = ZFFS_NULL;
+	dir->node.next = dir->list.head;
+
+	sys_slist_append(&zffs->opened, (void *)dir);
+
+	return 0;
+}
+
+int zffs_dir_read(struct zffs_data *zffs, struct zffs_area_pointer *pointer,
+		   struct zffs_dir *dir, struct zffs_node_data *data)
+{
+	int rc;
+	sys_snode_t *snode;
+
+	if (zffs_slist_is_tail(&dir->node)) {
+		return -ENOENT;
+	}
+
+	rc = zffs_slist_next(zffs, pointer, &dir->node);
+	if (rc < 0) {
+		return rc;
+	}
+
+	data->id = dir->node.id;
+	return zffs_misc_load_node(zffs, pointer, rc, data, &snode);
+}
+
+int zffs_dir_close(struct zffs_data *zffs,
+		    struct zffs_area_pointer *pointer, struct zffs_dir *dir)
+{
+	int rc = dir_save(zffs, pointer, dir);
+
+	if (rc) {
+		return rc;
+	}
+
+	sys_slist_find_and_remove(&zffs->opened, (void *)dir);
+
+	return 0;
+}
+
+// int zffs_dir_append(struct zffs_data *zffs,
```

Comment:
use /**/, not //

Reply:
Already not needed, forget to clean up when refactoring